modules/apps/27-interchain-accounts/types/codec_test.go

```diff
@@ -41,118 +44,291 @@ func (mockSdkMsg) GetSigners() []sdk.AccAddress {
+						//    that do not match the original message, causing equality checks to fail.
+						//
+						// 2. proto.Equal() does not have built-in support for comparing sdk's math.Int types.
+						//
+						// Using proto.CompactTextString() mitigates these issues by focusing on serialized string representation,
+						// rather than internal details of the types.
+						suite.Require().Equal(proto.CompactTextString(msg), proto.CompactTextString(deserializedMsgs[i]))
+					}
+				}
+			})
+		}
+
+		// test serializing non sdk.Msg type
+		bz, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []proto.Message{&banktypes.MsgSendResponse{}}, encoding)
+		suite.Require().NoError(err)
+		suite.Require().NotEmpty(bz)
+
+		// test deserializing unknown bytes
+		_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, bz, encoding)
+		suite.Require().Error(err) // unregistered type
+
+		// test deserializing unknown bytes
+		msgs, err := types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, []byte("invalid"), types.EncodingProtobuf)
+		suite.Require().Error(err)
+		suite.Require().Empty(msgs)
+	}
 }
 
 // unregistered bytes causes amino to panic.
 // test that DeserializeCosmosTx gracefully returns an error on
 // unsupported amino codec.
-func (suite *TypesTestSuite) TestDeserializeAndSerializeCosmosTxWithAmino() {
+func (suite *TypesTestSuite) TestProtoDeserializeAndSerializeCosmosTxWithAmino() {
 	cdc := codec.NewLegacyAmino()
 	marshaler := codec.NewAminoCodec(cdc)
 
-	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}})
+	msgs, err := types.SerializeCosmosTx(marshaler, []proto.Message{&banktypes.MsgSend{}}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(msgs)
 
-	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0})
+	bz, err := types.DeserializeCosmosTx(marshaler, []byte{0x10, 0}, types.EncodingProtobuf)
 	suite.Require().Error(err)
 	suite.Require().Empty(bz)
 }
+
+func (suite *TypesTestSuite) TestUnsupportedEncodingType() {
+	// Test serialize
+	msgs := []proto.Message{
+		&banktypes.MsgSend{
+			FromAddress: TestOwnerAddress,
+			ToAddress:   TestOwnerAddress,
+			Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+		},
+	}
+	_, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, msgs, "unsupported")
+	suite.Require().Error(err)
+
+	// Test deserialize
+	msgs = []proto.Message{
+		&banktypes.MsgSend{
+			FromAddress: TestOwnerAddress,
+			ToAddress:   TestOwnerAddress,
+			Amount:      sdk.NewCoins(sdk.NewCoin("bananas", sdkmath.NewInt(100))),
+		},
+	}
+	data, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, msgs, types.EncodingProtobuf)
+	suite.Require().NoError(err)
+	_, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, data, "unsupported")
+	suite.Require().Error(err)
```

Comment:
ditto: we can empty check the `[]sdk.Msg` too

```suggestion
	}
	
	data, err := types.SerializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, msgs, types.EncodingProtobuf)
	suite.Require().NoError(err)
	suite.Require().NotNil(data)
	
	msgs, err = types.DeserializeCosmosTx(simapp.MakeTestEncodingConfig().Marshaler, data, "unsupported")
	suite.Require().Error(err)
	suite.Require().Empty(msgs)
```

Reply:
done